Source/items.cpp

```diff
@@ -3331,32 +3336,23 @@ void TryRandomUniqueItem(Item &item, _item_indexes idx, int8_t mLevel, int uper,
 		return;
 	}
 
-	const UniqueItem &uniqueItem = UniqueItems[uid];
-	int targetLvl = 1; // Target level for reverse compatibility, since vanilla always takes the last applicable uid in the list.
-
-	// Set target level. Ideally we use uper 15 to have a 16% chance of generating a unique item.
-	if (uniqueItem.UIMinLvl - 4 > 0) { // Negative level will underflow. Lvl 0 items may have unintended consequences.
-		uper = 15;
-		targetLvl = uniqueItem.UIMinLvl - 4;
-	} else {
-		uper = 1;
-		targetLvl = uniqueItem.UIMinLvl;
+	// Find our own id to calculate the offset in validUniques
+	int uidOffset = -1;
+	for (size_t i = 0; i < validUniques.size(); i++) {
+		if (validUniques[i] == uid) {
+			// vanilla always picks the last unique, so the offset is calculated from the back of the valid unique list
+			uidOffset = static_cast<int>(validUniques.size() - i - 1);

```

Comment:
I added the generated logic back, but only use it when it can work (there is no unique with the same level after the desired unique).

Reply:
@kphoenix137 I'm marking these comments as resolved based on @obligaron's comments, let me know if you do not agree

Reply:
Trying to read over and understand the code. It's quite difficult because this whole concept is quite complicatated. :D

But from what I'm looking at here:
```
			// Found an item with same or lower level as our desired unique after our unique.
			// This means that we cannot possibly generate the item in reverse compatible mode and must rely on an offset.```
			
So I'm walking myself through the logic, say there's a valid uniques list of:
Item 1, ilvl 20 (i = 0)
Item 2, ilvl 21 (i = 1)
Item 3, ilvl 25 (i = 2)

And we have selected Item 2 (i = 1) as the desired unique on the initialization and assignment of `uid`, and the item level of the item generated before this function was called is at least ilvl 25.

The first loop here reaches the if-else because `0 != 1`, and assigns a uid offset of `3 - 0 - 1`, so a uid offset of 2.

Then we set `canGenerateReverseCompatible` to `false`.

Am I following this correctly? Also, we absolutely can generate this unique reverse compatible. That's where `targetLvl` comes into play. By force generating a new item with either ilvl 21 and `uper = 1`, or ilvl 17 and `uper = 15`, any unique item generated will always be our desired unique item.

Reply:
> So I'm walking myself through the logic, say there's a valid uniques list of: Item 1, ilvl 20 (i = 0) Item 2, ilvl 21 (i = 1) Item 3, ilvl 25 (i = 2)
> 
> And we have selected Item 2 (i = 1) as the desired unique on the initialization and assignment of `uid`, and the item level of the item generated before this function was called is at least ilvl 25.
> 
> The first loop here reaches the if-else because `0 != 1`, and assigns a uid offset of `3 - 0 - 1`, so a uid offset of 2.
> 
> Then we set `canGenerateReverseCompatible` to `false`.
> 
> Am I following this correctly?

The first iteration does nothing because Item 1 isn't the item we selected and `uidOffset` is still -1.

The second iteration sets `uidOffset = 3 - 1 - 1` because Item 2 is the item we selected. So we get a uid offset of 1, not 2.

The third iteration does nothing because Item 3 isn't the item we selected and ilvl 25 is bigger than ilvl 21.

So at the end, `canGenerateReverseCompatible` is `true` and we will go ahead and enter the loop to regenerate this item with a different ilvl.

Reply:
I see my mistake. Does this logic then match the desired logic where a uid offset is only utlized in situations where we truly cannot use a target level to get the desired uid?

Reply:
Yeah, it does match that. I think the main difference here is if the logic determines that a reverse-compatible item cannot be generated, it doesn't try to change the ilvl or anything. It just regenerates the same item with the appropriate UID offset. This means, in the worst case, that it may generate Bramble on an ilvl 11 ring with a UID offset of 12 in Hellfire. IIRC, the previous logic would have gone ahead and regenerated an ilvl 1 ring with a UID offset of 1.

The good news is that this worst-case scenario doesn't overrun the four bits you allocated for the offset. The bad news for modders is that it might be a bit less resilient to new items being added to the unique item data.

Reply:
Is there much of an advantage to not force generating with the desired target level when using uid offset? If I'm understanding you correctly, still generating a new unique with the target level is more future proofed for modders.